### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by Rufy86** [SVGKit/SVGKit#429]
I've tried to use the documentation linked in readme page. but result old and not updated. there is a page in which there is updated documentation and that explain not only the visualization of a SVG file but also the parsing a file, the animation and so on?

**Comment by adamgit** [comment]
Parsing and animation etc haven't changed in the past 4 years or so.

Can you provide a specific line of docs that is wrong?

**Comment by Rufy86** [comment]
For example; SVGKParser\* parser = [SVGKParser parserWithDefaultParseExtensions]; Error is:No known class method for selector 'parserWithDefaultParseExtensions'.

I'm trying to use this framework to get from SVG file some CoreAnimation layer. but, sorry, reading the documentation the only thing I can do is load the image in SVGKLayeredImageView and add the subview. 

This is the code I take reading your documentation. If I want load an SVG file, get something of CoreAnimation or else, and animate every level, how can I do it? for example, i want animate the stroke of KanjiVG's file, how can I do it?

` // Do any additional setup after loading the view, typically from a nib.
    SVGKImage *image = [SVGKImage imageNamed:@"kanji.svg"];
    SVGKLayeredImageView *imageView = [[SVGKLayeredImageView alloc]initWithSVGKImage:image];
    [imageView setFrame:CGRectMake(0, 40, 100, 100)];
    [self.view addSubview:imageView];

```
SVGKParser* parser = [SVGKParser parserWithDefaultParseExtensions];//Here there is the error. How can I use the parser? Is really useful? If you see the code under, code that i take from documentation, it seems that I can parse manually.


NSString* tagToFind = @"g";
NodeList* result = [image.DOMDocument getElementsByTagName:tagToFind];

for( Element* domElement in result ){
    SVGElement* startPoint = (SVGElement*) domElement;
    NSString* tagToFind = @"path";
    NodeList* result = [startPoint getElementsByTagName:tagToFind];
    for( Element* domElement in result ){

        SVGElement* svgElement = (SVGElement*) domElement; // if your tags are all supported by SVGKit, they will be converted to SVGElement instances
        CALayer* layerUserClickedOn = svgElement;

        if( [layerUserClickedOn isKindOfClass:[CAShapeLayer class]] ) // should ALWAYS be true, but just in case you write your code wrong...
        {

            CAShapeLayer* shapeLayer = (CAShapeLayer*) layerUserClickedOn;

            shapeLayer.fillColor = [UIColor redColor].CGColor;

        }
        NSLog(@"elemento");
    }

}`
```

**Comment by adamgit** [comment]
You only need to use the parser if you're writing a custom XML file format.

Look at the properties of SVGKImage - you will find the DOM tree there,
which has the getElementsByTagName methods etc.

**Comment by Rufy86** [comment]
there is a documentation that explain completely this classes?

because, if I've understand well, this code is 'correct', right?

Here i'm trying to color a stroke, but I can't. It color something and every stroke are grouped in one side.
repeat:there is a documentation that explain how to make this and also things? if I want for every stroke a CoreAnimation Layer or something else to animate, how can I do it?
`NSString\* tagToFind = @"g";
NodeList\* result = [image.DOMDocument getElementsByTagName:tagToFind];

for( Element\* domElement in result ){
    SVGElement\* startPoint = (SVGElement_) domElement;
    NSString_ tagToFind = @"path";
    NodeList\* result = [startPoint getElementsByTagName:tagToFind];
    for( Element\* domElement in result ){

```
    SVGElement* svgElement = (SVGElement*) domElement; // if your tags are all supported by SVGKit, they will be converted to SVGElement instances
    CALayer* layerUserClickedOn = svgElement;

    if( [layerUserClickedOn isKindOfClass:[CAShapeLayer class]] ) // should ALWAYS be true, but just in case you write your code wrong...
    {

        CAShapeLayer* shapeLayer = (CAShapeLayer*) layerUserClickedOn;

        shapeLayer.fillColor = [UIColor redColor].CGColor;

    }
    NSLog(@"elemento");
}
```

}`

**Comment by adamgit** [comment]
I think you need to read Apples docs first - you don't seem to understand 
CAShapeLayer

**Comment by Rufy86** [comment]
to tell the truth, i'm trying to use your documentation to do something, to learn how to use this framework. In few word, i done this, reading your documentation. For this reason i'm asking you if there is a complete documentation or tutorials that explain how to use correctly this framework. Because i think that It has great potential and if there were some tutorial that explain from basic to high level (using core text or else) it can be used more easily.

**Comment by adamgit** [comment]
Apple has provided many, many pages of very high-quality documentation on
CoreAnimation and QuartzCore.

It would take us years to recreate their docs, and there is no benefit:
Apple has already done this, it is all available free, and it's "official"
from Apple.

The people who volunteer for this project are NOT going to re-write docs
that already exist. If you only want to draw an SVG, you can do it with a
few lines of code given in the examples - or look at the sample project
that shows how to do it multiple different ways.

If you want to program with it, we assume that you will read the official
Apple docs first and learn how to use CA, QC, etc.

**Comment by Rufy86** [comment]
perhaps there is a misunderstanding. I apologize. i'm asking an other thing in which Apple doc about CA, QC has nothing to do with it. NodeList, NodeElement, DOMDocuments, Element, SVGElement and so on, are part of this framework. not Apple doc.Your documentation is not complete. is insufficient. First, I use your framework to get from SVG file a CAShapeLayer, CGPAth, CGBezierPath or what you want. After that I manage this Apple classes. But if I can't understand what there is in Element object, or how can I loop inside an standard multilevel SVG file and and understand how get single part in CA, QC classes, is not useful know how work Apple classes. Repeat:first transform SVG path in Apple Path, after manager Apple path. understand? this is the problem.
forgive me if I'm a little angry but I need to understand how to use this framework because I want to animate the strokes of kanji of KanjiVG's file. And because use core text i can do it with my glyph and my project can be great. Please,  help me to understand how to use this framework

**Comment by adamgit** [comment]
SVGKImage has properties for the DOM tree (you must read the DOM
specification - google "W3 DOM" for full docs), and for the CALayer tree
(you must read Apple docs - google "CALayer" for full docs).

There are examples linked to from the front page of SVGKit showing how you
_might_ use the DOM tree to select specific nodes, and how you _might_ then
change your rendering of the CALayers.

The only thing you need to know that's specific to SVGKit is:
1. When you change the DOM, there is no change to the layers until/unless
   you call "newCALayerTree" (method on SVGKImage)

If you want to change the layers directly, your original SVG file must have
"id" tags for each node you want to change, then you can find them inside
the trees (off the top of my head, there's an example of this on front page
- but people have been making changes to this recently so you might need to
  read the SVGKImage.h file and / or the SVGKImage.m file for details)

...but once you have the CAShapeLayer etc, or the CGPath, you're no longer
using SVGKit, you're using Apple's libraries.

**Comment by Rufy86** [comment]
Thank you for your answer. Now I understand a little more. Let me check if I really understand.

`<svg xmlns="http://www.w3.org/2000/svg" width="109" height="109" viewBox="0 0 109 109">
<g id="kvg:StrokePaths_08cea" style="fill:none;stroke:#000000;stroke-width:3;stroke-linecap:round;stroke-linejoin:round;">
<g id="kvg:08cea" kvg:element="質">
    <g id="kvg:08cea-g1" kvg:position="top">
        <g id="kvg:08cea-g2" kvg:element="斤" kvg:position="left">
            <path id="kvg:08cea-s1" kvg:type="㇒" d="M43.05,11.84c-0.05,1.04-0.57,1.92-1.2,2.53c-3,2.89-6.23,4.88-13.9,7.93"/>
            <path id="kvg:08cea-s2" kvg:type="㇒" d="M25.96,21.62c0.8,0.88,1.29,2.21,1.29,3.16c0.01,15.46-2.62,20.34-8.18,25.21"/>
            <path id="kvg:08cea-s3" kvg:type="㇐" d="M29.22,29.63c0.83-0.03,2.06-0.15,3.53-0.34c4.46-0.57,11.15-1.69,15.77-2.27c1.26-0.16,2.37-0.27,3.24-0.32"/>
            <path id="kvg:08cea-s4" kvg:type="㇑" d="M39.41,28.55c0.73,0.73,1.22,2.31,1.22,3.3c0,0.79-0.12,2.42-0.08,10.4c0.01,2.12,0.04,3.54,0.08,3.75"/>
        </g>
        <g id="kvg:08cea-g3" kvg:element="斤" kvg:position="right">
            <path id="kvg:08cea-s5" kvg:type="㇒" d="M76.8,11.09c0.14,0.93-0.08,1.75-0.79,2.42c-3.76,3.49-6.88,5.99-13.3,8.79"/>
            <path id="kvg:08cea-s6" kvg:type="㇒" d="M60.58,21.62c0.57,0.73,0.89,1.84,0.91,2.64c0.26,9.11-0.12,15.49-5.43,21.23"/>
            <path id="kvg:08cea-s7" kvg:type="㇐" d="M63.47,27.88c0.65-0.02,1.71-0.12,3.04-0.26c5.56-0.61,15.93-2.13,21.98-2.83c1.47-0.17,2.68-0.29,3.51-0.34"/>
            <path id="kvg:08cea-s8" kvg:type="㇑" d="M76.25,26.62c0.73,0.73,1.22,2.31,1.22,3.3c0,0.79-0.12,2.42-0.08,10.4c0.01,2.12,0.04,3.54,0.08,3.75"/>
        </g>
    </g>
    <g id="kvg:08cea-g4" kvg:element="貝" kvg:position="bottom" kvg:radical="general">
        <g id="kvg:08cea-g5" kvg:element="目" kvg:position="top">
            <path id="kvg:08cea-s9" kvg:type="㇑" d="M35.12,51.3c0.73,0.73,1.22,2.31,1.22,3.3c0,0.79-0.12,19.17-0.08,27.15c0.01,2.12,0.04,3.54,0.08,3.75"/>
            <path id="kvg:08cea-s10" kvg:type="㇕a" d="M37.08,53.11c6.04-0.48,29.35-2.67,32.43-2.87c3.74-0.24,4.37,1.26,4.37,4.37c0,4.14-0.06,15.94,0,24.89c0.01,1.69,0.03,3.23,0.06,4.5"/>
            <path id="kvg:08cea-s11" kvg:type="㇐a" d="M37.28,63.13C45.25,62.5,62,61,72.8,60.36"/>
            <path id="kvg:08cea-s12" kvg:type="㇐a" d="M37.55,72.8c11.2-1.05,24.2-2.05,34.85-2.75"/>
            <path id="kvg:08cea-s13" kvg:type="㇐a" d="M37.67,83.38c9.08-0.88,22.95-2,34.85-2.75"/>
        </g>
        <g id="kvg:08cea-g6" kvg:position="bottom">
            <path id="kvg:08cea-s14" kvg:type="㇒" d="M40.73,87.75c0.02,1-0.23,2.25-1.25,2.97c-3.08,2.16-11.58,6.73-17.73,9.28"/>
            <path id="kvg:08cea-s15" kvg:type="㇔" d="M66.5,87.5c7.74,2.98,16.89,8.39,19,11"/>
        </g>
    </g>
</g>
</g>
<g id="kvg:StrokeNumbers_08cea" style="font-size:8;fill:#808080">
    <text transform="matrix(1 0 0 1 35.75 10.50)">1</text>
    <text transform="matrix(1 0 0 1 19.50 29.50)">2</text>
    <text transform="matrix(1 0 0 1 38.50 25.48)">3</text>
    <text transform="matrix(1 0 0 1 33.50 38.50)">4</text>
    <text transform="matrix(1 0 0 1 68.50 10.50)">5</text>
    <text transform="matrix(1 0 0 1 54.75 29.75)">6</text>
    <text transform="matrix(1 0 0 1 76.50 23.68)">7</text>
    <text transform="matrix(1 0 0 1 69.75 36.50)">8</text>
    <text transform="matrix(1 0 0 1 27.75 58.03)">9</text>
    <text transform="matrix(1 0 0 1 43.20 50.45)">10</text>
    <text transform="matrix(1 0 0 1 43.20 60.88)">11</text>
    <text transform="matrix(1 0 0 1 43.20 69.80)">12</text>
    <text transform="matrix(1 0 0 1 43.20 80.23)">13</text>
    <text transform="matrix(1 0 0 1 25.50 92.50)">14</text>
    <text transform="matrix(1 0 0 1 56.70 92.50)">15</text>
</g>
</svg>
`

This is my SVG file.

in my code i hade, of course, load file
`SVGKImage *image = [SVGKImage imageNamed:@"kanji.svg"];` Right?

after that ? what I have to do?

The file, as you can see, you have more than one ID. image so I'll have to make a loop, right?

`NSString\* tagToFind = @"id";
    NodeList\* result = [image.DOMDocument getElementsByTagName:tagToFind];

```
for( Element* domElement in result ){
    SVGElement* startPoint = (SVGElement*) domElement;
```

//This SVG Element, what it contains? and what do I do? How can I turn it into a CAShaper Layer or in a CGPath?
            }
            NSLog(@"elemento");
        }

```
}`
```

This SVG Element, what it contains? and what do I do? How can I turn it into a CAShaper Layer or in a CGPath?

**Comment by adamgit** [comment]
For every element in the DOMtree that has an "id" tag, there is a
CAShapeLayer with the same "id" tag in the CALayerTree.

I can't remember how you get one from the other - there are convenience
methods to help you, or you can manually iterate over the whole tree
searching for them. The convenience methods are described in the "recipes"
linked from front page, and should not have changed significantly.

**Comment by Rufy86** [comment]
> there are convenience methods to help you

for the benefit of all, can you tell me what are these convenient methods?
after that, there are some big problem in your document:

in [http://t-machine.org/index.php/2013/01/02/svgkit-2013-recipes/](url)

there is this code:
`NSString* idInSVGFile = ... // assuming your SVG file has an "id" attribute for this node`  <-what does mean this comment? this string where I take it from?
`CALayer* absoluteLayer = [newImage newCopyPositionedAbsoluteLayerWithIdentifier:isInSVGFile];` //isInSVGFile I think is an error because should be **idInSVGFile**, right?

```

... //Here what should I do? Animation of CALayer absoluteLayer?

// NB: "absoluteLayer" is now positioned in absolute space;
//   if you add it to your window using e.g.: 
[self.view.layer addSublayer: absoluteLayer];
// ...it will appear in the same place as it appeared before,
//   keeping all the offsets, rotations, etc


```

Other errors  and incomprehensible piece

```

NSString* tagToFind = @"linearGradient";
NodeList* result = [svgImage.DOMDocument getElementsByTagName:tagToFind];

for( Element* domElement in result )
{
   // You can use the Element object directly:
   domElement.tagName; //Where I'm sending? And what return and what contains? This is useful for what?
   ...

   // ...or, if it was parsed by the SVG parser, you can convert it to an SVGElement:
   SVGElement* svgElement = (SVGElement*) domElement; //This is more correct and comprehensible.
   ...
}`

```

I'm sorry about this comment but show you that the documentation is sparse and does not help to use your framework. and a framework that big can not have documentation so little, or at least a lack of big tutorial that can help to understand the functioning of the framework classes.

For Example:
if I want Render one small part of the SVG file on its own ( is one of recipe), what i have to to? this?

```
SVGKImage *image = [SVGKImage imageNamed:@"kanji.svg"];
NSString* tagToFind = @"id";
    NodeList* result = [image.DOMDocument getElementsByTagName:tagToFind];

    for (Element *domElement in result) {
        NSString* idInSVGFile = ... // assuming your SVG file has an "id" attribute for this node //??????
CALayer* absoluteLayer = [newImage newCopyPositionedAbsoluteLayerWithIdentifier:isInSVGFile];
// Do you know if i write this code in Xcode, and It gets bored and spits out, rightly, errors?
    }
`

```

**Comment by Rufy86** [comment]
Continuing to try and using the imagination I could get something with the help of the Debug.

This is code:

```
SVGKImage *image = [SVGKImage imageNamed:@"kanji.svg"];
    SVGKLayeredImageView *imageView = [[SVGKLayeredImageView alloc]initWithSVGKImage:image];
    [imageView setFrame:CGRectMake(0, 40, 100, 100)];
    //[self.view addSubview:imageView];


    NSString* tagToFind = @"g";
    NodeList* result = [image.DOMDocument getElementsByTagName:tagToFind];


    for (Element *domElement in result) {
        SVGElement *startPoint = (SVGElement *)domElement;
        NSString *tagToFind = @"path";
        NodeList *resultBis = [startPoint getElementsByTagName:tagToFind];
        for (Element *domElementBis in resultBis) {
            SVGElement *svgElement = (SVGElement *)domElementBis;
            CALayer* layerUserClickedOn = (CALayer *)svgElement;
            CAShapeLayer* shapeLayer = (CAShapeLayer*) layerUserClickedOn;
            shapeLayer.strokeColor = [UIColor redColor].CGColor;
            [self.view.layer addSublayer: layerUserClickedOn];
        }   
    }
```

this code crashes because domElementBis is not SVGElement but is SVGPathElement, class not explained in documentation. And I, how do I turn this SVGPathElement into a CGPaths, CGShapeLayer or other class that I can do the animation? What do I do with a SVGPathElement?

I repet: The documentation is inadequate, old and useless.

**Comment by adamgit** [comment]
SVGPathElement is a subclass of SVGElement, by definition, so that error
must be wrong.

Note that SVGPathElement etc is all part of the SVG Specification - these
are fully documented on the W3 website. There are also direct links to the
docs for almost every such file in the header.

Re: "what do I do with a SVGPathElement" - if you want CA\* classes, you
don't want to use SVG\* classes - all you need is the "id" attribute, and
that will get you the CA_Layer from the CALayerTree, just as it will get
you the SVG_Element from the DOMTree.

Finally -- if you make a start at writing new tutorials, people here find
it easier to comment and help improve them than to write new ones
themslves. So it is a good way to get them: start them, and ask for
feedback :).

**Comment by Rufy86** [comment]
then, If I understand well, the problem is that I loop on DomTree and not of CALayerTree, is that right?

if so then my question is this: according to the documentation to search for items look for them through the tag. After that, I use this method that returns a NodeList which I suppose is an array type data structure, right?
`NodeList * result = [svgImage.DOMDocument getElementsByTagName: tagToFind];`
This array contains the SVGGElement.

how do I loop over CALayerTree?
There is  an similar method to get an NodeList full of CALayers? or how can I get an list of Layers? Can I transform SVGGElement in a CALayer?

I found the SVGKImage's method called 'dictionaryOfLayer' that returns a dictionary, but it gives me error saying that the key can not be nil. 

If I can figure how to do these things, I might do tutorials. but I need it myself understand how the elements.

**Comment by adamgit** [comment]
"Can I transform SVGGElement in a CALayer?"

-- yes, that's the intended approach. The idea is that you write the
smallest amount of code possible! DOM (the official standard) has these
built-in methods for "finding" things by tag, so we didn't want to
re-invent the wheel. We let you find the SVGElement in the DOM tree, then
use a method to automatically find the equivalent CALayer.

"I found the SVGKImage's method called 'dictionaryOfLayer' that returns a
dictionary, but it gives me error saying that the key can not be nil."

-- there were recent discussions about this, and some recent changes in
code. It had some bug(s) which I think are now fixed - but maybe they broke
it in new ways!

NB: the dictionary uses the value of the "id" tag as the key. Your SVG file
you copy/pasted has unique "id" tags for almost everything, so if you know
which id you want, you don't even need to use the DOM tree: request it
directly from the dictionaryOfLayers.

If you don't find what you're looking for, use the debugger to look at
dictionaryOfLayers and see what's in there - it SHOULD have one key/value
for each "id"-labelled tag in your original file.

**Comment by Rufy86** [comment]
```
yes, that's the intended approach. The idea is that you write the
smallest amount of code possible! DOM (the official standard) has these
built-in methods for "finding" things by tag, so we didn't want to
re-invent the wheel. We let you find the SVGElement in the DOM tree, then
use a method to automatically find the equivalent CALayer.

```

Which method let me transform an SVGElement, SVGPathElement in a CALayer?
other question: there is any method to transform this SVGKit Classes in CGPath, or CAShapeLayers, UIBezierPaths or something? If there is, which are this methods?

**Comment by adamgit** [comment]
I think you want to read from the dictionaryOfLayers ? If not, I don't
understand your question, sorry.

**Comment by Rufy86** [comment]
i've tried dictionaryOfLayers method because i was trying to find an structure of data that let me loop on CALevelTree. but you sad that the approach is to use method like
`NodeList * result = [svgImage.DOMDocument getElementsByTagName: tagToFind];`

to find part and after that to use method to transform it in CALayer, right? then I want to use this method. but i can't find the method to transform SCGGElement, SVGPathElement or what else in CALayer.

After that i've asked if there is some method to get, not CALayer but an CGPath o other I written before.

Sorry for my english.

**Comment by Rufy86** [comment]
however, I want to thank you for your patience and the help you're giving me. I am slowly understanding how it works and certainly once I figured, why not, I will do the tutorials that can help other people to use the framework

**Comment by adamgit** [comment]
Each volunteer focusses on what matters to them most (since none of us get
paid to work on SVGKit).

Personally, I focus on helping people add to the codebase, and keeping the
codebase clear, clean, and effective - and also: gradually pushing us
closer to 100% implementation of the SVG specification.

I've written quite a lot of docs, but I stopped doing that so I could focus
on the code things I just listed. I felt they were a bigger problem!

**Comment by Rufy86** [comment]
I see. it Is great thing.

About my question, can you answer? what is the method let me convert SCGGElement, SVGPathElement or what else in CALayer? And, there are method that let me convert SCGGElement, SVGPathElement or what else in in CGPath, or CAShapeLayers, UIBezierPaths or something like this?

I explain you. because i want to show how write an Kanji, I need take from KanjiVG' file every stroke and animate the draw. for example, if the stroke is a vertical line, I want to show an animation that display an line that is drawing from left to right. understand?

I don't know if I explain me well. if you don't understand the question, say me.

**Comment by adamgit** [comment]
I don't have access to a computer right now, so I can't give any detailed
adivce, I have to speak from memory.

I believe the dictionaryOfLayers is what you want. Use the debugger. See
what's in that dictionary.

**Comment by Rufy86** [comment]
Now I've an other problem, connect to documentation: I'm trying to add SVGKit in a new project and after to have add all framework i've this problem at runtime:
dyld: Library not loaded: @rpath/SVGKit.framework/SVGKit
  Referenced from: /var/containers/Bundle/Application/771F53E1-6ED6-4901-98F6-FB85A442C48C/KanjiAnimationTest.app/KanjiAnimationTest
  Reason: image not found

I've done every thing is written in Readme page to add the framework, but come out this.

come out also whether there is an image in project and loaded with SVGKImage and added in view with SVGKLayeredImageView.

How can I fix it?

**Comment by Rufy86** [comment]
I've solved problem about this problem: dyld: Library not loaded: @rpath/SVGKit.framework/SVGKit

In reame page is written:You MAY also need to add the following 3rd party Frameworks to your app (included with SVGKit):

CocoaLumberjack (in the SVGKit-iOS project, select the "3rd-party-frameworks" folder, and drag/drop whichever versions you need: iOS, tvOS, etc) Remember to embed the library too (should show under Target -> General -> Embedded Binaries)

But this is not only for Cocoa Lumberjack but also SVGKit framework. I didn't know this. now the problem is fixed.

**Comment by Rufy86** [comment]
About the using of dictionaryOfLayers the problem is crash at this method:
the method: `[layersByID setValue:layer forKey:layerID];` make crash because layerID is nil.
and how you can see there is an TODO.
`- (void) addSVGLayerTree:(CALayer_) layer withIdentifier:(NSString_) layerID toDictionary:(NSMutableDictionary*) layersByID
{
    // TODO: consider removing this method: it caches the lookup of individual items in the CALayerTree. It's a performance boost, but is it enough to be worthwhile?
    [layersByID setValue:layer forKey:layerID];

```
if ( [layer.sublayers count] < 1 )
{
    return;
}

for (CALayer *subLayer in layer.sublayers)
{
    NSString* subLayerID = [subLayer valueForKey:kSVGElementIdentifier];

    if( subLayerID != nil )
    {
        SVGKitLogVerbose(@"[%@] element id: %@ => layer: %@", [self class], subLayerID, subLayer);

        [self addSVGLayerTree:subLayer withIdentifier:subLayerID toDictionary:layersByID];

    }
}
```

}
`

**Comment by revolter** [comment]
Where is the documentation?

**Comment by adamgit** [comment]
Front page of the website, and the wiki, cover the basics, with links to 
more advanced usage.

If there's anything you feel is missing, why not start a wiki page for.it?

**Comment by revolter** [comment]
The website is not linked anywhere on GitHub. It should have been right in the repo's description.

Well, the README is the first thing most users read, and it doesn't have any `Getting started` guide. Personally, I like READMEs like [AFNetworking][1] or [DateTools][2] have, which you can easily skim or search for basic usage and then go to the documentation from the code for more advanced usages.

Also, the links in the wiki are pointing to some blog posts, which also have incorrect (old?) code examples.

I don't understand how so many people starred this but didn't help it. If I end up using this I might contribute to it to have a nicer welcome experience.

  [1]: https://github.com/AFNetworking/AFNetworking
  [2]: https://github.com/MatthewYork/DateTools

**Comment by adamgit** [comment]
When I joined the project there were many incompatible versions, and it
took us a couple of years (!)  to clean that up and get it all into one,
with the bugfixes from different branches, etc. So ... for a couple of
years ... we used the README as the docs, because that let us have
different parallel sets of docs for different versions. No need for that
any more - IMHO we should migrate most of the README into the Wiki!

NB: "The website" *is* github :).

I agree it would be great to have an updated Getting Started - I think the
examples were written for 1.2.x, just before we switched to 2.x.

Wiki would be the best place, with links from the front page, then
anyone/everyone can update (and has no excuse not to ;)).

**Comment by revolter** [comment]
Yeah, that makes sense.

I thought the website is [this](http://svgkit.sourceforge.net) one but didn't read anything on it 😆.

**Comment by adamgit** [comment]
That's an unrelated project that happens to have the same name :).

AIUI, SVGKit was named because Apple names their libraries "something"-Kit.

Coincidentally, for a while there was a trend for Javascript authors to do
the same thing.

**Comment by Rufy86** [comment]
@revolter As I sad, there is not any things that you can call seriously 'documentation'. There is some web page that explain something but you can not use it, because is OLD. and Xcode give you errors.
but the saddest thing is the answer:"If there's anything you feel is missing, why not start a wiki page for.it?".

I would like to help @adamgit but, I'm sorry! it's not my project. To create documentation is a work of owner of the project of their contributors. but not of us users. It's a rule of Software Engineering.

So. Please. stop you to insert new unuseful and very hard to use features and  make an updated, complete documentation. This project could be a great framework if was more simple and, above all, documented.

**Comment by revolter** [comment]
@Rufy86, Wrong! If you use this open source software, take some time and contribute to it! That way, you become a contributor too, so you won't have any excuse. What you said it's not a rule for OSS.

**Comment by justinclift** [comment]
> I would like to help @adamgit but, I'm sorry! it's not my project. To create documentation is a work of owner of the project of their contributors. but not of us users. It's a rule of Software Engineering.

Pretty much every Open Source Software project with a Community would disagree with you strongly.  The whole point of collaborative development is... collaboration to make things better for each other.  eg better code, better docs, better testing, etc.

**Comment by adamgit** [comment]
I have a lot of sympathy for @Rufy86 's argument here; I've encountered OSS projects before that had no docs, were hard to use, confusing, nonsensical - but no-one wanted to document them. It was very frustrating to me as a developer trying to use them

However:

"As I sad, there is not any things that you can call seriously 'documentation'."

...because it's not needed.

Most of the library is defined by the W3C spec, we have no choice in how it works or what it does. You can read the entire spec yourself - it's free, google "W3C SVG specification".

As for the rest ... this library has been very carefully written to be a very very simple "drag/drop/run". We copy/pasted Apple's own header files to make it identical to UIImage, and I personally wrote a lot of code that makes it do "the right thing" if you try to use it in an obvious way. e.g. SVGKImage.size does the right things, with a lot of clever code behind it - even if you change it to a new value.

Even the code architecture has been heavily designed to be simple and obvious:

1. Read file from somewhere
2. Parse file into DOM (as required by the spec!)
3. Export DOM into UIImage .. or into an SVG text file ... or into Apple's preferred CoreAnimation layers ... etc

All those things are done for you, so you don't need to know any of them, but if you run the debugger you'll see them happening, with obvious names, doing obvious things, in obvious ways. This makes it very easy to edit and make your changes.

The only bit that's complicated is when someone wants to add a missing feature of the SVG spec - then we tend to give them guidance on where we got stuck when we tried to do it ourselves, or why the code isn't working, or what we think is missing - so they can finish/fix it quickly.

So, in short: this project is one of the few that doesn't "need" docs, beyond simple install instructions. I'd still like to see them, but I'm not going to write them myself ... I HAVE ALREADY WRITTEN THEM and published them and put links up on the site ... if you want to update them, please go ahead!

**Comment by adamgit** [comment]
PS: to make it even easier to find the spec, I manually went through every header file that is defined by the W3C spec and found the URL for that file in the W3C website and copy/pasted it at the top of the header file.

So, when you're looking at SVGKit in Xcode, you can click through to read the docs directly!
